Subject: [PATCH] git-instaweb: bind to IPv4 and IPv6 host addresses

Note that gitweb.cgi does not properly handle literal IPv6
addresses, but at least with this change connecting to
localhost (or the host's actual name) will work properly
regardless of whether or not the connection is IPv4 or IPv6.

We now will attempt to connect to localhost instead dof
127.0.0.1 on a browse action if we configured an IPv6 address.

Signed-off-by: Kyle J. McKay <mackyle@gmail.com>

---
 git-instaweb.sh | 23 +++++++++++++++++++----
 1 file changed, 19 insertions(+), 4 deletions(-)

diff --git a/git-instaweb.sh b/git-instaweb.sh
index 6897e0b8..2c0c5223 100755
--- a/git-instaweb.sh
+++ b/git-instaweb.sh
@@ -30,6 +30,7 @@ root="$(git config --get instaweb.gitwebdir)"
 port=$(git config --get instaweb.port)
 module_path="$(git config --get instaweb.modulepath)"
 action="browse"
+hostname="127.0.0.1"
 
 conf="$GIT_DIR/gitweb/httpd.conf"
 
@@ -44,6 +45,16 @@ test -z "$root" && root='@@GITWEBDIR@@'
 # any untaken local port will do...
 test -z "$port" && port=1234
 
+select_hostname () {
+	case "$httpd" in
+	*lighttpd*)
+		# we always configure IPv6 support for lighttpd
+		# so connect to localhost instead of 127.0.0.1
+		hostname="localhost"
+		;;
+	esac
+}
+
 resolve_full_httpd () {
 	case "$httpd" in
 	*apache2*|*lighttpd*|*httpd*)
@@ -252,9 +263,13 @@ EOF
 }
 
 lighttpd_conf () {
+	host4="0.0.0.0"; host6="[::]"
+	test x"$local" = xtrue && { host4="127.0.0.1"; host6="[::1]"; }
 	cat > "$conf" <<EOF
 server.document-root = "$root"
 server.port = $port
+server.bind = "$host4"
+\$SERVER["socket"] == "$host6:$port" {}
 server.modules = ( "mod_setenv", "mod_cgi" )
 server.indexfiles = ( "gitweb.cgi" )
 server.pid-file = "$fqgitdir/pid"
@@ -325,7 +340,6 @@ mimetype.assign             = (
   ""              =>      "text/plain"
  )
 EOF
-	test x"$local" = xtrue && echo 'server.bind = "127.0.0.1"' >> "$conf"
 }
 
 apache2_conf () {
@@ -694,10 +708,11 @@ conf="$fqgitdir/gitweb/$httpd_only.conf"
 configure_httpd
 
 start_httpd
-url=http://127.0.0.1:$port
+select_hostname
+url="http://$hostname:$port"
 
 if test -n "$browser"; then
-	httpd_is_ready && git web--browse -b "$browser" $url || echo $url
+	httpd_is_ready && git web--browse -b "$browser" "$url" || echo "$url"
 else
-	httpd_is_ready && git web--browse -c "instaweb.browser" $url || echo $url
+	httpd_is_ready && git web--browse -c "instaweb.browser" "$url" || echo "$url"
 fi
---
